CelesteNet.Client/Components/CelesteNetPico8Component.cs

```diff
@@ -0,0 +1,467 @@
+            _emulatorData = null;
+            _classicData = null;
+            _classic = null;
+        }
+        
+        _emulatorData ??= DynamicData.For(emu);
+        _classic ??= (Classic?) _emulatorData.Get("game");
+        if (_classic != null)
+            _classicData ??= DynamicData.For(_classic);
+
+        return !(_classic == null || _emulatorData == null || _classicData == null);
+    }
+
+    private bool _inGame;
+
+    public override void Update(GameTime gameTime) {
+        base.Update(gameTime);
+
+        if (Engine.Scene is not Emulator) {
+            Ghosts.Clear();
+            return;
+        }
+
+        if (!Settings.Connected) return;
+        if (!_alive) {
+            Ghosts.Clear();
+            return;
+        }
+
+        lock (Ghosts) {
+            foreach (var kvp in Ghosts)
+                PruneIfInactive(kvp.Key, kvp.Value);
+        }
+
+        if (!InitData()) return;
+
+        var objs = (List<Classic.ClassicObject>?) _classicData?.Get("objects");
+        if (objs == null) return;
+
+        lock (objs) {
+            if (!objs.Any(o => o is Classic.player || o is Classic.player_spawn)) {
+                Logger.Log(LogLevel.DBG, "PICO8-CNET", $"NO PLAYER FOUND");
+                _alive = false;
+                return;
+            }
+        }
+
+        var state = new DataPicoFrame {
+            Player = Client?.PlayerInfo,
+            Spr = _queuedGhostFrame.Spr,
+            X = _queuedGhostFrame.X,
+            Y = _queuedGhostFrame.Y,
+            FlipX = _queuedGhostFrame.FlipX,
+            FlipY = _queuedGhostFrame.FlipY,
+            Type = _queuedGhostFrame.Type,
+            DJump = _queuedGhostFrame.DJump,
+            Hair = _queuedGhostFrame.Hair,
+            Level = LevelIndex,
+        };
+
+        Logger.Log(LogLevel.DBG, "PICO8-CNET", $"SEND {state}");
+
+        Client?.Send(state);
+    }
+
+    private void PruneIfInactive(uint id, PicoGhost ghost)
+    {
+        bool active = 
+            Settings.Connected &&
+            _inGame &&
+            ghost.Player.ID == id &&
+            (Client?.Data?.TryGetBoundRef(ghost.Player, out DataPlayerState? state) ?? false) &&
+            state is { SID: "PICO-8" } &&
+            LevelIndex != -1 &&
+            state.Level == $"{(LevelIndex + 1) * 100}M";
+            // TODO: need more checks!
+        
+        if (!active) Ghosts.TryRemove(id, out _);
+    }
+
+    private PicoGhost? InitGhost(DataPlayerInfo player) {
+        if (Engine.Scene is not Emulator emu) {
+            Logger.Log(LogLevel.WRN, "PICO8-CNET", "Tried to initialize pico ghost in non-Emulator");
+            return null;
+        }
+        
+        if (!InitData()) {
+            Logger.Log(LogLevel.WRN, "PICO8-CNET", "Failed to InitData");
+            return null;
+        }
+
+        if (_classic == null) {
+            Logger.Log(LogLevel.WRN, "PICO8-CNET", "Failed to retrieve Classic");
+            return null;
+        }
```

Comment:
if _classic was null, InitData wouldn't have returned true and the previous conditional would've returned, right